File: leetcode/medium/122_maxProfit.py
from typing import List
import json
from collections import deque, defaultdict

class Solution:
    def maxProfit_0(self, prices: List[int]) -> int:
        expenses = [0]*len(prices)
        expenses[0] = prices[0]
        income = [0]*len(prices)

        for i in range(1, len(prices)):
            expenses[i] = min(expenses[i-1], prices[i], prices[i]-income[i-1])
            income[i] = max(income[i-1], prices[i] - expenses[i-1])
        return income[-1]

'''
7,1,5,3,6,4
7 1 1-1-1-3
0 0 4 4 7 7

expenses[i] = min(expenses[i-1], prices[i], prices[i]-income[i-1])
income[i] = max(income[i-1], prices[i] - expenses[i-1])

'''

def test ():
    params = [
        {
            'input': [7,1,5,3,6,4],
            'output': 7,
        },
        {
            'input': [1,2,3,4,5],
            'output': 4,
        },
        {
            'input': [7,6,4,3,1],
            'output': 0,
        },
    ]
    solution = Solution()

    for param in params:
        result = solution.maxProfit(param['input'])
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()

File: leetcode/medium/122_maxProfit.ts
function maxProfit(prices: number[]): number {
    const costs: number[] = new Array(prices.length).fill(0);
    costs[0] = prices[0];
    const gain: number[] = new Array(prices.length).fill(0);

    for (let i: number = 1; i < prices.length; ++i) {
        costs[i] = Math.min(costs[i-1], prices[i] - gain[i-1]);
        gain[i] = Math.max(gain[i-1], prices[i] - costs[i]);
    }

    return gain.at(-1) || 0;
};

/*

p   [7, 1, 5, 3, 6, 4]
c   [7, 1, 1,-1,-1,-3]
g   [0, 0, 4, 4, 7, 7]

costs[n] = min(costs[n-1], price[n] - gain[n-1])
gain[n] = max(gain[n-1], price[n] - costs[n])


*/



const test = () => {
    const params = [
        {
            input: [7,1,5,3,6,4],
            output: 7,
        },
        {
            input: [1,2,3,4,5],
            output: 4,
        },
        {
            input: [7,6,4,3,1],
            output: 0,
        },
    ];

    params.forEach(({input, output}) => {
        const result = maxProfit(input);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();